Fix BLE "Notifications already enabled" race condition

Dinges28 opened the pull request:
Ignore harmless BleakError when notifications are already active to prevent log spam.


When using the Idasen Desk integration together with ESP32 Bluetooth Proxy, Desk.connect() may attempt to re-enable BLE notifications that are already active. This causes Bleak to raise:

BleakError: Notifications are already enabled


This exception is harmless but currently propagates, leading to task crashes and continuous log spam.

This PR catches this specific error in IdasenDeskCoordinator.async_connect() and logs it on debug level instead, while still re-raising all other exceptions.

This prevents log flooding and stabilizes reconnect behavior without masking real connection issues.

Tested with multiple Idasen desks via ESP32 Bluetooth Proxy.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [ ] I understand the code I am submitting and can explain how it works.
- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.
- [ ] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Dinges28 committed:
Fix BLE "Notifications already enabled" race condition

Ignore harmless BleakError when notifications are already active to prevent log spam.


When using the Idasen Desk integration together with ESP32 Bluetooth Proxy,
Desk.connect() may attempt to re-enable BLE notifications that are already active.
This causes Bleak to raise:

BleakError: Notifications are already enabled


This exception is harmless but currently propagates, leading to task crashes and
continuous log spam.

This PR catches this specific error in IdasenDeskCoordinator.async_connect() and
logs it on debug level instead, while still re-raising all other exceptions.

This prevents log flooding and stabilizes reconnect behavior without masking real
connection issues.

Tested with multiple Idasen desks via ESP32 Bluetooth Proxy.

Dinges28 committed:
Merge branch 'dev' into patch-1

abmantis commented:
> > Thanks for opening the PR @Dinges28 I haven't yet looked too deeply into what causes this log, but I think we should fix the cause of the warning, probably in the lib idasen-ha, instead of suppressing it.
> > Are you able to consistently reproduce this and test it?
> 
> Yes, I can consistently test it.. it floods my logs ;-) I now solve it by disconnecting the desks and reconnecting when necessary

Would it fix it to ensure that it is disconnected before calling connect again?

Dinges28 commented:
> > > Thanks for opening the PR @Dinges28 I haven't yet looked too deeply into what causes this log, but I think we should fix the cause of the warning, probably in the lib idasen-ha, instead of suppressing it.
> > > Are you able to consistently reproduce this and test it?
> > 
> > 
> > Yes, I can consistently test it.. it floods my logs ;-) I now solve it by disconnecting the desks and reconnecting when necessary
> 
> Would it fix it to ensure that it is disconnected before calling connect again?

I think that it will fix it.. It gives the error that it is already connected, connecting again is than not an expected command

Dinges28 committed:
Fix root cause: prevent reconnect while already connected

As far as my knowledge goes and a little help of ChatGPT I came up with the following pull request

This fixes the root cause instead of suppressing the error.
The coordinator now tracks connection state and prevents reconnect attempts while the desk is already connected.
This removes the BLE race condition where notifications are re-enabled while still active, which caused the “Notifications already enabled” BleakError.

Dinges28 commented:
> I assumed you would implement the disconnect -> connect fix in the library! If not, I'll get to it eventually (it was already on my TODO list).

I edited my PR and to my knowledge I troubleshooted the issue.

Dinges28 committed:
Merge branch 'dev' into patch-1

abmantis reviewed (changes requested):
This should be implemented in [idasen-ha](https://github.com/abmantis/idasen-ha), since the connection logic is handled there.

abmantis reviewed (commented):
Closing this since it was fixed by https://github.com/abmantis/idasen-ha/pull/37 and will be updated in https://github.com/home-assistant/core/pull/163309
